host/sgx/ocalls.c

```diff
@@ -41,28 +41,26 @@ void HandleThreadWait(oe_enclave_t* enclave, uint64_t arg_in)
 
 #if defined(__linux__)
 
-    if (__sync_fetch_and_add(&event->value, (uint32_t)-1) == 0)
+    if (__sync_fetch_and_add(event, (uint32_t)-1) == 0)
     {
         do
         {
-            syscall(
-                __NR_futex,
-                &event->value,
-                FUTEX_WAIT_PRIVATE,
-                -1,
-                NULL,
-                NULL,
-                0);
+            syscall(__NR_futex, event, FUTEX_WAIT_PRIVATE, -1, NULL, NULL, 0);
             // If event->value is still -1, then this is a spurious-wake.
             // Spurious-wakes are ignored by going back to FUTEX_WAIT.
             // Since FUTEX_WAIT uses atomic instructions to load event->value,
             // it is safe to use a non-atomic operation here.
-        } while (event->value == (uint32_t)-1);
+        } while (*event == (uint32_t)-1);
     }
 
 #elif defined(_WIN32)
 
-    WaitForSingleObject(event->handle, INFINITE);
+    int32_t zero = 0;
+    int32_t one = 1;
+    if (_InterlockedCompareExchange(event, one, zero) == one)
+    {
+        WaitOnAddress(event, &one, sizeof(event), INFINITE);
```

Comment:
It has to be `sizeof(*event)` which would be 4 bytes instead of 8.

Reply:
Can you double check the logic?

As written it seems to be:

```c
    if (_InterlockedCompareExchange(event, one, zero) == one)  <- if event is one, then set it to zero
    {
        WaitOnAddress(event, &one, sizeof(event), INFINITE);     <- Wait until event is not one? This would return immediately?
```

Reply:
> 
> 
> It has to be `sizeof(*event)` which would be 4 bytes instead of 8.

Thanks, yes sizeof(*event) is right.

Reply:
You are right again there is some error. Change to if (_InterlockedCompareExchange(event, one, zero) == zero).